File: apps/api/src/agreements/agreements.service.ts
import { BadRequestException, ForbiddenException, Injectable, NotFoundException } from "@nestjs/common";
import { createHash, randomUUID } from "node:crypto";
import { PrismaService } from "../common/prisma/prisma.service";
import { AuthenticatedUser } from "../common/decorators/current-user.decorator";
import { renderDefaultAgreement } from "./agreement-template";
import { NotificationsService } from "../notifications/notifications.service";

const SIGNING_LINK_TTL_DAYS = 14;

@Injectable()
export class AgreementsService {
  constructor(
    private readonly prisma: PrismaService,
    private readonly notificationsService: NotificationsService
  ) {}

  /**
   * Generates a tenancy agreement for a confirmed booking and issues one
   * authenticated signing link per signatory (tenant + the property's
   * organisation). Refuses to regenerate over an already-signed agreement —
   * "no silent post-signing replacement" (Part I) is enforced here, not just
   * documented.
   */
  async generate(user: AuthenticatedUser, bookingId: string) {
    const booking = await this.prisma.booking.findUnique({
      where: { id: bookingId },
      include: {
        user: { include: { profile: true } },
        unit: {
          include: {
            property: { include: { organisation: true, houseRules: true } },
          },
        },
        agreements: true,
      },
    });
    if (!booking) {
      throw new NotFoundException("Booking not found.");
    }
    if (booking.userId !== user.userId && !user.roles.includes("Admin")) {
      throw new ForbiddenException("This booking belongs to a different account.");
    }
    if (booking.status !== "CONFIRMED") {
      throw new BadRequestException(
        `Cannot generate an agreement for a booking in status "${booking.status}"; it must be CONFIRMED (payment complete) first.`
      );
    }

    const alreadySealed = booking.agreements.some(
      (a: { status: string }) => a.status === "FULLY_SIGNED"
    );
    if (alreadySealed) {
      throw new BadRequestException(
        "This booking already has a fully signed agreement. Sealed agreements are never silently replaced — issue a formal amendment instead."
      );
    }

    const tenantName =
      [booking.user.profile?.firstName, booking.user.profile?.lastName].filter(Boolean).join(" ") ||
      booking.user.email ||
      booking.user.phone ||
      "Tenant";

    const bodyMarkdown = renderDefaultAgreement({
      tenantName,
      managerOrganisationName: booking.unit.property.organisation.name,
      propertyTitle: booking.unit.property.title,
      propertyAddress: booking.unit.property.address ?? "Information Required",
      unitLabel: booking.unit.publicLabel ?? "Information Required",
      agreedRent: `KES ${Number(booking.agreedRent).toLocaleString()}`,
      agreedDeposit: booking.agreedDeposit
        ? `KES ${Number(booking.agreedDeposit).toLocaleString()}`
        : "Information Required",
      moveInDate: booking.moveInDate ? booking.moveInDate.toDateString() : "Information Required",
      houseRules: booking.unit.property.houseRules,
    });

    const documentHash = createHash("sha256").update(bodyMarkdown).digest("hex");

    const agreement = await this.prisma.agreement.create({
      data: {
        bookingId: booking.id,
        status: "SENT",
        versions: {
          create: {
            version: 1,
            documentHash,
            bodyStorageKey: `agreements/${booking.id}/v1.md`, // real object storage is Milestone 4's media-pipeline follow-up
          },
        },
      },
      include: { versions: true },
    });

    const signatories = await Promise.all([
      this.prisma.agreementSignatory.create({
        data: { agreementId: agreement.id, userId: booking.userId, role: "tenant" },
      }),
      this.prisma.agreementSignatory.create({
        data: { agreementId: agreement.id, role: "manager" },
      }),
    ]);

    const signatureRequests = await Promise.all(
      signatories.map((signatory) =>
        this.prisma.signatureRequest.create({
          data: {
            signatoryId: signatory.id,
            authenticatedLinkToken: randomUUID(),
            expiresAt: new Date(Date.now() + SIGNING_LINK_TTL_DAYS * 24 * 60 * 60 * 1000),
          },
        })
      )
    );

    const tenantSigningLink = signatureRequests[0].authenticatedLinkToken;
    // Only the tenant signatory has a real userId today — the manager
    // signatory is created role-only (see the module report: resolving
    // "which specific manager user should sign" is a follow-up, not yet
    // implemented), so only the tenant gets notified here.
    await this.notificationsService.notify({
      userId: booking.userId,
      type: "agreement_signing_link",
      subject: "Your tenancy agreement is ready to sign",
      body: `Please review and sign your tenancy agreement for ${booking.unit.property.title}: /agreements/sign/${tenantSigningLink}`,
      payload: { agreementId: agreement.id },
    });

    return {
      agreement,
      bodyMarkdown,
      signingLinks: signatories.map((s, i) => ({
        role: s.role,
        token: signatureRequests[i].authenticatedLinkToken,
      })),
    };
  }

  /**
   * Fetches an agreement by its per-signatory link token. Never trusts a
   * booking or agreement id directly from an unauthenticated caller — the
   * token IS the authentication, per Part I ("authenticated signing links").
   */
  async getByToken(token: string, ipAddress?: string) {
    const request = await this.prisma.signatureRequest.findUnique({
      where: { authenticatedLinkToken: token },
      include: {
        signatory: {
          include: {
            agreement: { include: { versions: { orderBy: { version: "desc" }, take: 1 } } },
          },
        },
      },
    });
    if (!request) {
      throw new NotFoundException("Signing link not found or already used.");
    }
    if (request.expiresAt < new Date()) {
      throw new BadRequestException("This signing link has expired.");
    }

    await this.prisma.signatureEvent.create({
      data: { signatureRequestId: request.id, eventType: "viewed", ipAddress },
    });

    return {
      role: request.signatory.role,
      alreadySigned: Boolean(request.signatory.signedAt),
      agreementStatus: request.signatory.agreement.status,
      latestVersion: request.signatory.agreement.versions[0],
    };
  }

  /**
   * Records consent and seals the signature. Once every signatory on the
   * agreement has signed, the agreement moves to FULLY_SIGNED and a
   * SignedDocument is sealed — after which `generate()` above refuses to
   * ever touch this agreement again.
   */
  async sign(token: string, ipAddress?: string) {
    const request = await this.prisma.signatureRequest.findUnique({
      where: { authenticatedLinkToken: token },
      include: { signatory: { include: { agreement: { include: { signatories: true } } } } },
    });
    if (!request) {
      throw new NotFoundException("Signing link not found.");
    }
    if (request.expiresAt < new Date()) {
      throw new BadRequestException("This signing link has expired.");
    }
    if (request.signatory.signedAt) {
      throw new BadRequestException("This signatory has already signed.");
    }

    await this.prisma.signatureEvent.create({
      data: { signatureRequestId: request.id, eventType: "consented", ipAddress },
    });
    await this.prisma.signatureEvent.create({
      data: { signatureRequestId: request.id, eventType: "signed", ipAddress },
    });

    await this.prisma.agreementSignatory.update({
      where: { id: request.signatory.id },
      data: { signedAt: new Date() },
    });

    const allSignatories = await this.prisma.agreementSignatory.findMany({
      where: { agreementId: request.signatory.agreementId },
    });
    const stillPending = allSignatories.filter(
      (s: { id: string; signedAt: Date | null }) => s.id !== request.signatory.id && !s.signedAt
    );

    if (stillPending.length === 0) {
      await this.prisma.agreement.update({
        where: { id: request.signatory.agreementId },
        data: { status: "FULLY_SIGNED" },
      });
      await this.prisma.signedDocument.create({
        data: {
          agreementId: request.signatory.agreementId,
          storageKey: `agreements/${request.signatory.agreementId}/sealed.pdf`,
        },
      });
      return { fullySigned: true };
    }

    await this.prisma.agreement.update({
      where: { id: request.signatory.agreementId },
      data: { status: "PARTIALLY_SIGNED" },
    });
    return { fullySigned: false };
  }
}

File: apps/api-py/app/services/agreements_service.py
"""Direct port of agreements/agreements.service.ts."""

import hashlib
import uuid
from datetime import datetime, timedelta, timezone

from fastapi import HTTPException
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.deps import AuthenticatedUser
from app.models import (
    Agreement,
    AgreementSignatory,
    AgreementVersion,
    Booking,
    HouseRule,
    Organisation,
    Property,
    SignatureEvent,
    SignatureRequest,
    SignedDocument,
    Unit,
    User,
    UserProfile,
)
from app.services import notifications_service
from app.services.agreement_template import render_default_agreement

SIGNING_LINK_TTL_DAYS = 14


async def generate(db: AsyncSession, user: AuthenticatedUser, booking_id: str) -> dict:
    """Generates a tenancy agreement for a confirmed booking and issues one
    authenticated signing link per signatory (tenant + the property's
    organisation). Refuses to regenerate over an already-signed
    agreement — enforced here, not just documented."""
    booking = (await db.execute(select(Booking).where(Booking.id == booking_id))).scalar_one_or_none()
    if booking is None:
        raise HTTPException(status_code=404, detail="Booking not found.")
    if booking.user_id != user.user_id and "Admin" not in user.roles:
        raise HTTPException(status_code=403, detail="This booking belongs to a different account.")
    if booking.status != "CONFIRMED":
        raise HTTPException(
            status_code=400,
            detail=f'Cannot generate an agreement for a booking in status "{booking.status}"; it must be CONFIRMED (payment complete) first.',
        )

    existing_agreements = (
        await db.execute(select(Agreement).where(Agreement.booking_id == booking_id))
    ).scalars().all()
    if any(a.status == "FULLY_SIGNED" for a in existing_agreements):
        raise HTTPException(
            status_code=400,
            detail="This booking already has a fully signed agreement. Sealed agreements are never silently replaced — issue a formal amendment instead.",
        )

    unit = (await db.execute(select(Unit).where(Unit.id == booking.unit_id))).scalar_one()
    prop = (await db.execute(select(Property).where(Property.id == unit.property_id))).scalar_one()
    organisation = (await db.execute(select(Organisation).where(Organisation.id == prop.organisation_id))).scalar_one()
    house_rules = (await db.execute(select(HouseRule).where(HouseRule.property_id == prop.id))).scalars().all()
    booking_user = (await db.execute(select(User).where(User.id == booking.user_id))).scalar_one()
    profile = (
        await db.execute(select(UserProfile).where(UserProfile.user_id == booking.user_id))
    ).scalar_one_or_none()

    tenant_name = (
        " ".join(filter(None, [profile.first_name if profile else None, profile.last_name if profile else None])).strip()
        or booking_user.email
        or booking_user.phone
        or "Tenant"
    )

    body_markdown = render_default_agreement(
        {
            "tenant_name": tenant_name,
            "manager_organisation_name": organisation.name,
            "property_title": prop.title,
            "property_address": prop.address or "Information Required",
            "unit_label": unit.public_label or "Information Required",
            "agreed_rent": f"KES {float(booking.agreed_rent):,.0f}",
            "agreed_deposit": f"KES {float(booking.agreed_deposit):,.0f}" if booking.agreed_deposit else "Information Required",
            "move_in_date": booking.move_in_date.strftime("%a %b %d %Y") if booking.move_in_date else "Information Required",
            "house_rules": [{"rule_type": r.rule_type, "detail": r.detail} for r in house_rules],
        }
    )

    document_hash = hashlib.sha256(body_markdown.encode()).hexdigest()

    agreement = Agreement(booking_id=booking.id, status="SENT")
    db.add(agreement)
    await db.flush()

    db.add(
        AgreementVersion(
            agreement_id=agreement.id,
            version=1,
            document_hash=document_hash,
            # Real object storage is a follow-up, matching the original's
            # own documented scope for this exact field.
            body_storage_key=f"agreements/{booking.id}/v1.md",
        )
    )

    tenant_signatory = AgreementSignatory(agreement_id=agreement.id, user_id=booking.user_id, role="tenant")
    manager_signatory = AgreementSignatory(agreement_id=agreement.id, role="manager")
    db.add(tenant_signatory)
    db.add(manager_signatory)
    await db.flush()

    tenant_request = SignatureRequest(
        signatory_id=tenant_signatory.id,
        authenticated_link_token=uuid.uuid4().hex,
        expires_at=datetime.now(timezone.utc) + timedelta(days=SIGNING_LINK_TTL_DAYS),
    )
    manager_request = SignatureRequest(
        signatory_id=manager_signatory.id,
        authenticated_link_token=uuid.uuid4().hex,
        expires_at=datetime.now(timezone.utc) + timedelta(days=SIGNING_LINK_TTL_DAYS),
    )
    db.add(tenant_request)
    db.add(manager_request)
    await db.commit()
    await db.refresh(agreement)

    # Only the tenant signatory has a real user_id today — the manager
    # signatory is created role-only (resolving "which specific manager
    # user should sign" is a follow-up, not yet implemented), so only the
    # tenant gets notified here.
    await notifications_service.notify(
        db,
        booking.user_id,
        "agreement_signing_link",
        "Your tenancy agreement is ready to sign",
        f"Please review and sign your tenancy agreement for {prop.title}: /agreements/sign/{tenant_request.authenticated_link_token}",
        {"agreementId": agreement.id},
    )

    return {
        "agreement": {"id": agreement.id, "bookingId": agreement.booking_id, "status": agreement.status},
        "bodyMarkdown": body_markdown,
        "signingLinks": [
            {"role": "tenant", "token": tenant_request.authenticated_link_token},
            {"role": "manager", "token": manager_request.authenticated_link_token},
        ],
    }


async def get_by_token(db: AsyncSession, token: str, ip_address: str | None) -> dict:
    """Fetches an agreement by its per-signatory link token. Never trusts
    a booking or agreement id directly from an unauthenticated caller —
    the token IS the authentication."""
    request = (
        await db.execute(select(SignatureRequest).where(SignatureRequest.authenticated_link_token == token))
    ).scalar_one_or_none()
    if request is None:
        raise HTTPException(status_code=404, detail="Signing link not found or already used.")

    now = datetime.now(request.expires_at.tzinfo) if request.expires_at.tzinfo else datetime.now()
    if request.expires_at < now:
        raise HTTPException(status_code=400, detail="This signing link has expired.")

    signatory = (
        await db.execute(select(AgreementSignatory).where(AgreementSignatory.id == request.signatory_id))
    ).scalar_one()
    agreement = (await db.execute(select(Agreement).where(Agreement.id == signatory.agreement_id))).scalar_one()
    latest_version = (
        await db.execute(
            select(AgreementVersion)
            .where(AgreementVersion.agreement_id == agreement.id)
            .order_by(AgreementVersion.version.desc())
        )
    ).scalars().first()

    db.add(SignatureEvent(signature_request_id=request.id, event_type="viewed", ip_address=ip_address))
    await db.commit()

    return {
        "role": signatory.role,
        "alreadySigned": bool(signatory.signed_at),
        "agreementStatus": agreement.status,
        "latestVersion": {"version": latest_version.version, "documentHash": latest_version.document_hash} if latest_version else None,
    }


async def sign(db: AsyncSession, token: str, ip_address: str | None) -> dict:
    """Records consent and seals the signature. Once every signatory has
    signed, the agreement moves to FULLY_SIGNED and a SignedDocument is
    sealed — after which generate() above refuses to ever touch this
    agreement again."""
    request = (
        await db.execute(select(SignatureRequest).where(SignatureRequest.authenticated_link_token == token))
    ).scalar_one_or_none()
    if request is None:
        raise HTTPException(status_code=404, detail="Signing link not found.")

    now = datetime.now(request.expires_at.tzinfo) if request.expires_at.tzinfo else datetime.now()
    if request.expires_at < now:
        raise HTTPException(status_code=400, detail="This signing link has expired.")

    signatory = (
        await db.execute(select(AgreementSignatory).where(AgreementSignatory.id == request.signatory_id))
    ).scalar_one()
    if signatory.signed_at:
        raise HTTPException(status_code=400, detail="This signatory has already signed.")

    db.add(SignatureEvent(signature_request_id=request.id, event_type="consented", ip_address=ip_address))
    db.add(SignatureEvent(signature_request_id=request.id, event_type="signed", ip_address=ip_address))
    signatory.signed_at = datetime.now(timezone.utc)
    await db.commit()

    all_signatories = (
        await db.execute(select(AgreementSignatory).where(AgreementSignatory.agreement_id == signatory.agreement_id))
    ).scalars().all()
    still_pending = [s for s in all_signatories if s.id != signatory.id and not s.signed_at]

    agreement = (await db.execute(select(Agreement).where(Agreement.id == signatory.agreement_id))).scalar_one()

    if len(still_pending) == 0:
        agreement.status = "FULLY_SIGNED"
        db.add(SignedDocument(agreement_id=signatory.agreement_id, storage_key=f"agreements/{signatory.agreement_id}/sealed.pdf"))
        await db.commit()
        return {"fullySigned": True}

    agreement.status = "PARTIALLY_SIGNED"
    await db.commit()
    return {"fullySigned": False}
